core/service/src/engine/threshold/service/resharer.rs

```diff
@@ -43,9 +44,240 @@ use crate::{
+        bucket = bucket_from_domain(&parsed_url_and_bucket)?;
+        let url = s3_bucket_url
+            .replace(&(bucket.clone() + "."), "")
+            .trim_end_matches('/')
+            .to_owned();
+        tracing::info!(s3_bucket_url, url, bucket, "Bucket from domain");
+        Ok((url, bucket))
+    } else {
+        let url = s3_bucket_url
+            .replace(&bucket, "")
+            .trim_end_matches('/')
+            .to_owned();
+        tracing::info!(s3_bucket_url, url, bucket, "Parsed S3 url");
+        Ok((url, bucket))
+    }
+}
+
+async fn fetch_public_materials_from_peers<
+    PubS: Storage + Send + Sync + 'static,
+    PrivS: Storage + Send + Sync + 'static,
+    G: ReadOnlyS3StorageGetter<R>,
+    R: StorageReader,
+>(
+    crypto_storage: &ThresholdCryptoMaterialStorage<PubS, PrivS>,
+    key_id: RequestId,
+    context_id: &ContextId,
+    key_digests: &HashMap<PubDataType, Vec<u8>>,
+    ro_storage_getter: &G,
+) -> anyhow::Result<FhePubKeySet> {
+    // fetch the context info
+    let context = {
+        let priv_storage = crypto_storage.get_private_storage();
+        let guard_storage = priv_storage.lock().await;
+        read_context_at_id(&(*guard_storage), context_id).await?
+    };
+    for node in context.mpc_nodes {
+        // so simplify logic, it's ok to iterate over myself too
+        //
+        // the public storage URL consists of the bucket name and the URL
+        // we need to parse this information accordingly
+        let (url, bucket) = split_url(&node.public_storage_url)?;
+
+        // this is not an operation that is frequently used, so we can create a new s3 client each time
+        let s3_client = build_anonymous_s3_client(Some(url::Url::parse(&url)?)).await?;
+        let pub_storage = ro_storage_getter.get_storage(
+            s3_client,
+            bucket,
+            None,
+            StorageType::PUB,
+            Some(Role::indexed_from_one(node.party_id as usize)),
+            None,
+        )?;
+
+        // attempt to fetch the public materials from this node
+        let public_key = read_pk_at_request_id(&pub_storage, &key_id)
+            .await
+            .map(|pk| match pk {
+                WrappedPublicKeyOwned::Compact(compact_pk) => compact_pk,
+            });
+
+        let server_key = read_versioned_at_request_id::<_, ServerKey>(
+            &pub_storage,
+            &key_id,
+            &PubDataType::ServerKey.to_string(),
+        )
+        .await;
+
+        match (public_key, server_key) {
+            (Ok(public_key), Ok(server_key)) => {
+                let actual_server_key_digest =
+                    safe_serialize_hash_element_versioned(&DSEP_PUBDATA_KEY, &server_key)?;
+                let actual_public_key_digest =
+                    safe_serialize_hash_element_versioned(&DSEP_PUBDATA_KEY, &public_key)?;
+
+                // verify the digests
+                let expected_public_key_digest = key_digests
```

Comment:
Nitpick: Since this is an argument for the function I would check this in the beginning of the function to avoid the overhead of setting up an S3 connection then to realize there is an issue with the input